src/precompile.c

```diff
@@ -88,6 +96,13 @@ void jl_write_compiler_output(void)
                            (const char*)s->buf, (size_t)s->size);
         }
     }
+    for (size_t i = 0; i < jl_current_modules.size; i += 2) {
+        if (jl_current_modules.table[i + 1] != HT_NOTFOUND) {
+            jl_printf(JL_STDERR, "\nWARNING: detected unclosed module: ");
+            jl_static_show(JL_STDERR, (jl_value_t*)jl_current_modules.table[i]);
+            jl_printf(JL_STDERR, "\n  ** incremental compilation may be broken for this module **\n\n");
+        }
+    }
```

Comment:
example output:
```
$ ./julia --output-ji x.ji --sysimage usr/lib/julia/sys.so -e 'module A; end; @async(@eval module As; sleep(1); end); yield(); module B; module C; exit(0); end; end' --output-incremental=yes

WARNING: detected unclosed module: B
  ** incremental compilation may be broken for this module **


WARNING: detected unclosed module: B.C
  ** incremental compilation may be broken for this module **


WARNING: detected unclosed module: As
  ** incremental compilation may be broken for this module **
```
I assume this never happens in real code (and thus isn't worthwhile to try to handle), but since this already added the infrastructure to be able detect it, might as well print a warning.